scalding-core/src/main/scala/com/twitter/scalding/Execution.scala

```diff
@@ -852,11 +870,22 @@ object Execution {
       { case (conf, _, w, cec) => w.getForced(conf, t)(cec) })
 
   def toIterable[T](t: TypedPipe[T]): Execution[Iterable[T]] =
+    from(UUID.randomUUID()).flatMap { uuid =>
+      toIterable(t, uuid)
+    }
+
+  def toIterableWithClose[T](t: TypedPipe[T]): Execution[(Iterable[T], Execution[Unit])] =
+    from(UUID.randomUUID()).flatMap { uuid =>
+      toIterable(t, uuid).map(WithCleanup(uuid)(_))
```

Comment:
you need a function for the `flatMap` too in order to make sure `Execution` caching works.

Note, you are defeating your case class with `WithCleanup(uuid)(_)`. It is already a function, you should write: `WithCleanup(uuid)`

The test to add is: `Execution.toIterableWithClose(a) == Execution.toIterableWithClose(a)`. I think we will fail this test at the moment due to the closures.

Reply:
done